src/entities/pools/gyro2/gyro2Pool.ts

```diff
@@ -0,0 +1,299 @@
+    }
+
+    public increase(amount: bigint): TokenAmount {
+        this.amount = this.amount + amount;
+        this.scale18 = this.amount * this.scalar;
+        return this;
+    }
+
+    public decrease(amount: bigint): TokenAmount {
+        this.amount = this.amount - amount;
+        this.scale18 = this.amount * this.scalar;
+        return this;
+    }
+}
+
+export class Gyro2Pool implements BasePool {
+    public readonly chainId: number;
+    public readonly id: Hex;
+    public readonly address: string;
+    public readonly poolType: PoolType = PoolType.Gyro2;
+    public readonly poolTypeVersion: number;
+    public readonly swapFee: bigint;
+    public readonly tokens: Gyro2PoolToken[];
+
+    private readonly sqrtAlpha: bigint;
+    private readonly sqrtBeta: bigint;
+    private readonly tokenMap: Map<string, Gyro2PoolToken>;
+
+    static fromRawPool(chainId: number, pool: RawGyro2Pool): Gyro2Pool {
+        const poolTokens: Gyro2PoolToken[] = [];
+
+        for (const t of pool.tokens) {
+            const token = new Token(
+                chainId,
+                t.address,
+                t.decimals,
+                t.symbol,
+                t.name,
+            );
+            const tokenAmount = TokenAmount.fromHumanAmount(token, t.balance);
+
+            poolTokens.push(
+                new Gyro2PoolToken(token, tokenAmount.amount, t.index),
+            );
+        }
+
+        return new Gyro2Pool(
+            pool.id,
+            pool.poolTypeVersion,
+            parseEther(pool.swapFee),
+            parseEther(pool.sqrtAlpha),
+            parseEther(pool.sqrtBeta),
+            poolTokens,
+        );
+    }
+
+    constructor(
+        id: Hex,
+        poolTypeVersion: number,
+        swapFee: bigint,
+        sqrtAlpha: bigint,
+        sqrtBeta: bigint,
+        tokens: Gyro2PoolToken[],
+    ) {
+        this.chainId = tokens[0].token.chainId;
+        this.id = id;
+        this.poolTypeVersion = poolTypeVersion;
+        this.swapFee = swapFee;
+        this.sqrtAlpha = sqrtAlpha;
+        this.sqrtBeta = sqrtBeta;
+        this.address = getPoolAddress(id);
+        this.tokens = tokens;
+        this.tokenMap = new Map(
+            this.tokens.map((token) => [token.token.address, token]),
+        );
+    }
+
+    public getNormalizedLiquidity(tokenIn: Token, tokenOut: Token): bigint {
+        const tIn = this.tokenMap.get(tokenIn.wrapped);
+        const tOut = this.tokenMap.get(tokenOut.wrapped);
+
+        if (!tIn || !tOut)
+            throw new Error('Pool does not contain the tokens provided');
+        // TODO: Fix gyro normalized liquidity calc
+        return tOut.amount;
+    }
+
+    public swapGivenIn(
+        tokenIn: Token,
+        tokenOut: Token,
+        swapAmount: TokenAmount,
+        mutateBalances?: boolean,
+    ): TokenAmount {
+        // TODO: try/catch is being used on previous SOR - should we keep the same bevaiour or should we throw an error?
+        try {
```

Comment:
No blanket try / catches - we should understand any failure scenarios and have detailed throws